skbio/metadata/_interval.py

```diff
@@ -0,0 +1,334 @@
+# ----------------------------------------------------------------------------
+
+from ._intersection import IntervalTree
+
+
+class Interval():
+    '''Stores the position and metadata of an interval.
+
+    Parameters
+    ----------
+    intervals : list of tuple of ints
+        List of tuples representing start and end coordinates.
+    boundaries : list of tuple of bool
+        List of tuples, representing the openness of each interval.
+    metadata : dict
+        Dictionary of attributes storing information of the feature
+        such as `strand`, `gene_name` or `product`.
+    interval_metadata : object
+        A reference to the `IntervalMetadata` object that this
+        interval is associated to.
+    '''
+    def __init__(self, intervals=None, boundaries=None,
+                 metadata=None, interval_metadata=None):
+        iv = []
+        for interval in intervals:
+            iv.append(_polish_interval(interval))
+
+        if boundaries is not None:
+            self.boundaries = boundaries
+        else:
+            self.boundaries = []
+
+        if metadata is not None:
+            self.metadata = metadata
+        else:
+            self.metadata = {}
+
+        self._interval_metadata = interval_metadata
+
+        if intervals is not None:
+            self.intervals = sorted(iv)
+        else:
+            self.intervals = []
+
+        if interval_metadata is not None:
+            self._add_interval()
+
+    def _add_interval(self):
+        """ Add an interval object to the IntervalTree within
+            IntervalMetadata."""
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in self.intervals:
+            if loc is not None:
+                start, end = loc
+                self._interval_metadata._intervals.add(start, end, self)
+        self._interval_metadata._metadata.append(self)
+
+    def __getitem__(self, key):
+        return self.metadata[key]
+
+    def __setitem__(self, key, val):
+        self.metadata[key] = val
+
+    # This is required for creating unique sets of intervals
+    def __hash__(self):
+        return hash(tuple(sorted(self.metadata.items()) +
+                          self.intervals +
+                          self.boundaries))
+
+    def __lt__(self, other):
+        return self.intervals < other.intervals
+
+    def __gt__(self, other):
+        return self.intervals > other.intervals
+
+    def __le__(self, other):
+        return self.intervals <= other.intervals
+
+    def __ge__(self, other):
+        return self.intervals >= other.intervals
+
+    def __eq__(self, other):
+        return ((self.metadata == other.metadata) and
+                (self.intervals == other.intervals) and
+                (self.boundaries == other.boundaries))
+
+    def __contains__(self, key):
+        return key in self.metadata
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __repr__(self):
+        return ''.join([
```

Comment:
👍 it seems really useful to include start/end. Your example repr sort of implies the interval stretches from 0 to 7, though it's non-contiguous within that. We could indicate `start=0`, `end=7`, and somehow indicate that it's non-contiguous, but maybe it's worth just repr-ing sorted `intervals` and including that in `Interval` repr:

```python
<Interval: intervals=[(0, 2), (4, 7)], 2 metadata keys, dropped=False>
```

If you don't want to show all intervals, you could truncate to show the first and last N intervals.

Here's another idea, using start/end:

```python
<Interval: start=0, end=7, 2 non-contiguous intervals, 2 metadata keys, dropped=False>
<Interval: start=4, end=9, 1 contiguous interval, 4 metadata keys, dropped=False>
```

I don't like including the `IntervalMetadata` memory address in the repr because it makes the repr unusable in doctests. From a user's standpoint, they're going to care more whether the `Interval` is in a dropped state or not, because that affects the public API they can access, and `dropped` is part of the public API as well.